### PR TITLE
LPS-108239 DB Partitioning

### PR DESCRIPTION
**Pull request by achaparro**
Hi @Preston-Crary,

This is the final implementation for DB Partitioning. Here you have the [complete definition of this new feature](https://docs.google.com/document/d/1sh7iwGaWhSAmf3ilRsMNljU8K0jbyEcbk3DsqEMCpTg/). 

Please, before reviewing it, take into account the following things:
- For now, the implementation is only valid for MySQL. We validate this on startup, see https://github.com/Preston-Crary/liferay-portal/pull/970/commits/bcc4591d4aa656ecde717395a3bc50f1dbe3fab3
- After this pull, I will remove the companyId from Portlet and I will the companyId to CTMessage and LayoutClassedModelUsage. Then we will be able to simplify this method: https://github.com/Preston-Crary/liferay-portal/pull/970/commits/76e9791c914b65ad674190c557afa4f60d90d963#diff-6bb24d9f61cea3f431b595701d675e89R136 For the future, we already have a new SF rule to require the companyId in new non-control tables.
- The Company removal would be handled en a separate JVM in Saas, for now we just avoid to execute the whole logic, see https://github.com/Preston-Crary/liferay-portal/commit/76e9791c914b65ad674190c557afa4f60d90d963#diff-6bb24d9f61cea3f431b595701d675e89R95
- After this pull, I will send a new one to execute the processes which don't use the CompanThreadLocal once per company such as:
  - Upgrade processes
  - Listeners (except model listeners and a few others)
  - Verifiers
  - Osgi commands which check the database

Thanks!

### CODE REVIEW

**Comment by achaparro** [comment]
ci:test:sf

**Comment by achaparro** [comment]
ci:test

File: portal-impl/src/com/liferay/portal/util/PropsValues.java [hunk 1]
```diff
@@ -606,6 +606,10 @@
 	public static final String DATABASE_MYSQL_ENGINE = PropsUtil.get(
 		PropsKeys.DATABASE_MYSQL_ENGINE);
 
+	public static final boolean DATABASE_PARTITIONING_ENABLED =
```
Comment (Preston-Crary) [inline portal-impl/src/com/liferay/portal/util/PropsValues.java]: Let's avoid adding the PropsValues/PropsKeys and just get the property where it is used.

File: portal-kernel/src/com/liferay/db/partitioning/kernel/util/DBPartitioningUtil.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+	private static final boolean _DATABASE_PARTITIONING_ENABLED =
+		GetterUtil.getBoolean(
+			PropsUtil.get(PropsKeys.DATABASE_PARTITIONING_ENABLED));
+
+	private static volatile DBPartitioning _dbPartitioning;
+
+	static {
+		if (_DATABASE_PARTITIONING_ENABLED) {
+			_dbPartitioning = ServiceProxyFactory.newServiceTrackedInstance(
```
Comment (Preston-Crary) [inline portal-kernel/src/com/liferay/db/partitioning/kernel/util/DBPartitioningUtil.java]: Let's always assign the _dbPartitioning field and check the property inside DBPartitioning.

File: portal-kernel/src/com/liferay/db/partitioning/kernel/util/DBPartitioning.java [hunk 3]
```diff
@@ -12,12 +12,33 @@
...
+
+import org.osgi.annotation.versioning.ProviderType;
+
+/**
+ * @author Alberto Chaparro
+ */
+@ProviderType
+public interface DBPartitioning {
+
+	public default void addPartition(long companyId) throws Exception {
```
Comment (Preston-Crary) [inline portal-kernel/src/com/liferay/db/partitioning/kernel/util/DBPartitioning.java]: Let's remove the default methods here. Just check the property in the DBPartitioningImpl.

File: modules/apps/static/portal/portal-db-partitioning-test/src/testIntegration/java/com/liferay/portal/db/partitioning/internal/test/DBPartitioningTest.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+/**
+ * @author Alberto Chaparro
+ */
+@RunWith(Arquillian.class)
+public class DBPartitioningTest {
+
+	@ClassRule
+	@Rule
+	public static final AggregateTestRule aggregateTestRule =
+		new LiferayIntegrationTestRule();
```
Comment (Preston-Crary) [inline modules/apps/static/portal/portal-db-partitioning-test/src/testIntegration/java/com/liferay/portal/db/partitioning/internal/test/DBPartitioningTest.java]: Let's use the AssumeTestRule here instead of all the DBType checks. Just have the static assume method.

File: modules/apps/static/portal/portal-db-partitioning-test/src/testIntegration/java/com/liferay/portal/db/partitioning/internal/test/DBPartitioningTest.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+			CompanyThreadLocal.setCompanyIdInitialization(currentCompanyId);
+
+			_dbPartitioning.usePartition(_connection);
+		}
+	}
+
+	@Test
+	public void testValidate() {
+		try {
+			_dbPartitioning.validate();
```
Comment (Preston-Crary) [inline modules/apps/static/portal/portal-db-partitioning-test/src/testIntegration/java/com/liferay/portal/db/partitioning/internal/test/DBPartitioningTest.java]: Let's set the DBType so this can always run and fail after the validate() call.

File: portal-impl/src/com/liferay/portal/service/impl/CompanyLocalServiceImpl.java [hunk 6]
```diff
@@ -1246,6 +1176,96 @@ public void updateSecurity(
 	protected void addAssetEntriesFacet(SearchContext searchContext) {
 	}
 
+	protected User addDefaultUser(Company company) {
```
Comment (Preston-Crary) [inline portal-impl/src/com/liferay/portal/service/impl/CompanyLocalServiceImpl.java]: Can you extract these methods in separate commits? Also let's make them private.

**Comment by Preston-Crary** [comment]
I don't see anything that requires MySQL only, do we really need to check this at all?

File: portal-kernel/src/com/liferay/db/partitioning/kernel/util/DBPartitioningUtil.java (hunk 2 above)
Comment (achaparro) [reply portal-kernel/src/com/liferay/db/partitioning/kernel/util/DBPartitioningUtil.java]: Hey @Preston-Crary, if we do that we make the portal depend upon the DB Partitioning module. We should avoid it since this is an internal functionality and we should allow customers to work without the module.

**Comment by achaparro** [comment]
The command "USE" is only valid in MySQL:
https://github.com/Preston-Crary/liferay-portal/pull/970/commits/76e9791c914b65ad674190c557afa4f60d90d963#diff-6bb24d9f61cea3f431b595701d675e89R117

There are other equivalents for most of the databases but we should validate and test it. Apart from this, here we create other schemas based on the default schema:
https://github.com/Preston-Crary/liferay-portal/pull/970/commits/76e9791c914b65ad674190c557afa4f60d90d963#diff-6bb24d9f61cea3f431b595701d675e89R47

Probably other database vendor like Oracle don't allow to create tables based on other database tables.

Anyway, it would be better to focus on one database vendor for now (since it's just an internal functionality) and support others in the future once this fully works.

Regarding the rest of the things, I will change the code based on your suggestions. The only thing I don't agree is this one:
https://github.com/Preston-Crary/liferay-portal/pull/970#discussion_r380867262

Thanks.

**Comment by Preston-Crary** [comment]
Let's move DBPartitioning to a better package and rename to DBPartitionHelper. com.liferay.portal.kernel.dao.db.partition

**Comment by achaparro** [comment]
I will work in the changes and I will send it again soon
